Feature/google pay integration

avargic opened the pull request.

avargic committed:
- initial google pay flow implementation
- implements startGooglePayPayment method and modifies confirmPayment
- adds google play activity to example app
- refactor monriHttpApiImpl to re-use code segments

avargic committed:
- addGooglePaySessionManager and startGooglePayment to get data before initializing the google pay button

avargic committed:
- minor changes

avargic committed:
- minor changes

avargic committed:
- move google pay metadata to app side

avargic committed:
- move PaymentDataRequest creation to helper

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+    }
+
+    private void processCreatePaymentSessionResponse(final NewPaymentResponse response) {
+        if (!response.getStatus().equals("approved")) {
```
"approved" u konstantu i radije okreni uvjete u if-u, ovako je malo zbunjujuće (source: ja san krivo pročita jer san očekiva obratno)

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+
+    private void processCreatePaymentSessionResponse(final NewPaymentResponse response) {
+        if (!response.getStatus().equals("approved")) {
+            resultTextView.setText(String.format("Create payment session rejected: %s", response.getStatus()));
```
Prebaci ove stringove koje koristis u `resultTextView` u resurse radije.

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+            @Override
+            public void onSuccess(final JSONObject googlePaySessionParameters) {
+                monriGooglePaymentRequestHelper.setMonriGooglePaySessionParameters(googlePaySessionParameters);
+                monriGooglePaymentRequestHelper.setTotalPriceLabel("Total price");
```
Konstanta

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+            @Override
+            public void onError(final Throwable throwable) {
+                progressBar.setVisibility(View.GONE);
+                resultTextView.setText(String.format("Start google pay payment error: %s", throwable));
```
Što se ispiše kao rezultat `throwable.toString()`? Je li isto što i `throwable.getMessage()`?

avargic replied:
mislim da je mala razlika, toString mislim da ce vratiti i ime/vrstu exceptiona i error message, a getMessage vraca samo taj error message

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+                                .addOnSuccessListener(this::showGooglePayButton)
+                                .addOnFailureListener(this::processOnReadyToPayFailure);
+        } else {
+            resultTextView.setText("Error: isReadyToPayRequest is null");
```
Ako se dogodi `JSONException` error poruka će biti vidljiva 1ms jer će se odmah prebrisati sa ovom. Ako je moguće da `getIsReadyToPayRequest()` vrati `null` onda radije prebaci if else u try blok

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+        try {
+            paymentDataRequest = monriGooglePaymentRequestHelper.getPaymentDataRequest();
+        } catch (JSONException e) {
+            throw new RuntimeException(e);
```
Ovo mi se čini kao da ne bi trebalo biti tu

luka-dragutin-asee commented on app/src/main/java/com/monri/android/example/GooglePayActivity.java:
```diff
@@ -0,0 +1,274 @@
+        final String paymentInfo = paymentData.toJson();
+
+        try {
+            googlePaymentMethodData = new JSONObject(paymentInfo).getJSONObject("paymentMethodData");
```
Isto u konstantu

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/MonriApiImpl.java:
```diff
@@ -20,6 +22,20 @@ class MonriApiImpl implements MonriApi {
+    public void startGooglePayPayment(final String paymentId, final ResultCallback<JSONObject> callback) {
+        taskRunner.executeAsync(
+                () -> {
+                    MonriHttpResult<JSONObject> result = monriHttpApi.startGooglePayPayment(paymentId);
```
final

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/MonriHttpApi.java:
```diff
@@ -28,6 +30,8 @@ static MonriHttpApi create(final String baseUrl, final Map<String, String> heade
     //post v2/payment/{id}/confirm
     MonriHttpResult<ConfirmPaymentResponse> confirmPayment(ConfirmPaymentParams confirmPaymentParams);
 
+    MonriHttpResult<JSONObject> startGooglePayPayment(final String PaymentId);
```
u interface ne treba final, samo kad se implementira metoda

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/MonriHttpApiImpl.java:
```diff
@@ -447,23 +262,29 @@ public static JSONObject confirmPaymentParamsToJSON(@NonNull ConfirmPaymentParam
         final Map<String, String> data = paymentMethodParams.getData();
 
-        JSONObject dataMapJSON = new JSONObject();
+        JSONObject paymentMethodJSON = new JSONObject();
```
final

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/MonriHttpApiImpl.java:
```diff
@@ -271,6 +69,23 @@ public MonriHttpResult<ConfirmPaymentResponse> confirmPayment(@NonNull ConfirmPa
+
     //get v2/payment/{id}/status
     @Override
     public MonriHttpResult<PaymentStatusResponse> paymentStatus(String id) {
```
final

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/MonriHttpApiImpl.java:
```diff
@@ -447,23 +262,29 @@ public static JSONObject confirmPaymentParamsToJSON(@NonNull ConfirmPaymentParam
-        JSONObject paymentMethodJSON = new JSONObject();
+            paymentMethodJSON.put(PAYMENT_METHOD_DATA_KEY, tokenDataObject);
+        } else {
+            JSONObject dataMapJSON = new JSONObject();
```
final

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/google_pay/MonriGooglePaymentRequestHelper.java:
```diff
@@ -0,0 +1,63 @@
+
+    public PaymentDataRequest getPaymentDataRequest() throws JSONException {
+
+        JSONObject paymentDataRequest = getBaseRequest()
```
final, i ispod isto

luka-dragutin-asee commented on monri/src/main/java/com/monri/android/model/GooglePayPayment.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public Map<String, String> data() {
+        Map<String, String> data = new HashMap<>();
```
final 😆

avargic committed:
- add alternative google pay flow to sdk
- fix MR comments

avargic committed:
- move strings to resources file

avargic committed:
- monri sdk version bump

avargic committed:
- target api version bump
- fix edge-to-edge display for confirmPayment activity

avargic committed:
- fix build tools and compileSdkVersion

avargic committed:
- fix final modifiers
- remove apiVersion and apiVersionMinor from MonriGooglePaymentRequestHelper constructor
- remove unused GooglePayHandler exception
- add Status model
- gradle version bump

avargic committed:
Merge pull request #47 from MonriPayments/feature/googlePayAlternativeFlow

Feature/google pay alternative flow
